server/saved_objects/saved_objects_wrapper.ts

```diff
@@ -0,0 +1,230 @@
+import { SecurityPluginConfigType } from '..';
+import {
+  DEFAULT_TENANT,
+  globalTenantName,
+  GLOBAL_TENANT_SYMBOL,
+  isPrivateTenant,
+  PRIVATE_TENANT_SYMBOL,
+} from '../../public/apps/configuration/utils/tenant-utils';
+import { OpenSearchDashboardsAuthState } from '../auth/types/authentication_type';
+
+export class SecuritySavedObjectsClientWrapper {
+  public httpStart?: HttpServiceStart;
+  public config?: SecurityPluginConfigType;
+
+  constructor() {}
+
+  public wrapperFactory: SavedObjectsClientWrapperFactory = (wrapperOptions) => {
+    const state: OpenSearchDashboardsAuthState =
+      (this.httpStart!.auth.get(wrapperOptions.request).state as OpenSearchDashboardsAuthState) ||
+      {};
+
+    const selectedTenant = state.selectedTenant;
+    const username = state.authInfo?.user_name;
+    const isGlobalEnabled = this.config!.multitenancy.tenants.enable_global;
+    const isPrivateEnabled = this.config!.multitenancy.tenants.enable_private;
+
+    let namespaceValue = selectedTenant;
+
+    const createWithNamespace = async <T = unknown>(
+      type: string,
+      attributes: T,
+      options?: SavedObjectsCreateOptions
+    ) => {
+      if (this.isAPrivateTenant(selectedTenant, isPrivateEnabled)) {
+        namespaceValue = selectedTenant! + username;
+      }
+      _.assign(options, { namespace: [namespaceValue] });
+      return await wrapperOptions.client.create(type, attributes, options);
+    };
+
+    const bulkGetWithNamespace = async <T = unknown>(
+      objects: SavedObjectsBulkGetObject[] = [],
+      options: SavedObjectsBaseOptions = {}
+    ): Promise<SavedObjectsBulkResponse<T>> => {
+      if (this.isAPrivateTenant(selectedTenant, isPrivateEnabled)) {
+        namespaceValue = selectedTenant! + username;
+      }
+      _.assign(options, { namespace: [namespaceValue] });
+      return await wrapperOptions.client.bulkGet(objects, options);
+    };
+
+    const findWithNamespace = async <T = unknown>(
+      options: SavedObjectsFindOptions
+    ): Promise<SavedObjectsFindResponse<T>> => {
+      const tenants = state.authInfo?.tenants;
+      const availableTenantNames = Object.keys(tenants!);
+      availableTenantNames.push(DEFAULT_TENANT);
+      if (isGlobalEnabled) {
+        availableTenantNames.push(GLOBAL_TENANT_SYMBOL);
+      }
+      if (isPrivateEnabled) {
+        availableTenantNames.push(PRIVATE_TENANT_SYMBOL + username);
+      }
+      if (availableTenantNames.includes(globalTenantName)) {
+        let index = availableTenantNames.indexOf(globalTenantName);
+        if (index > -1) {
+          availableTenantNames.splice(index, 1);
+        }
+        index = availableTenantNames.indexOf(username!);
+        if (index > -1) {
+          availableTenantNames.splice(index, 1);
+        }
+      }
+      const typeToNamespacesMap: any = {};
+      if (isPrivateTenant(selectedTenant!)) {
+        namespaceValue = selectedTenant! + username;
+      }
```

Comment:
@zengyan-amazon suggested to create a separate method to handle this logic for reuse.

Reply:
done